### PR TITLE
Document ARM64 unwind handling for return address signing

### PR DESCRIPTION
**Pull request by mstorsjo**
This is deduced/guessed from studying the output of MSVC with the `/d2guardsignret` flag. Hopefully someone with insight into the unwinder internals can confirm whether this indeed is the case.

I made up an unwind opcode name for this opcode (0xfc, `sign_ra`) - other suggestions are welcome.

I'm not entirely sure about whether this is how it is handled for the packed functions, but this would intuitively be how it works.

Since there are two parallel instructions for signing return addresses, `paciasp` and `pacibsp`, the unwinder would need to know which one to use for signing - MSVC seems to generate `pacibsp` (togeter with `autibsp`) - is there any other similar opcode for `paciasp`, or is that one unsupported for Windows unwinding purposes?

### CODE REVIEW

**Comment by corob-msft** [comment]
@mstorsjo 

Thanks for the proposal. As near as I can tell, the actual opcode is known as `pac_sign_return_address`, and the logic is largely as you've divined. There appears to be a flag that lets the signing/unsigning opcodes be replaced with nops, so there's an additional bit of complexity involved somewhere. At least, I think so, if I'm reading things correctly, which is always a question. The signing step is "Step 1" in the internal comments, and saving registers is "Step 2", so we should probably renumber those and the following steps accordingly.
 
To get the inside story, I'm invoking @laurenprinn and @sigatrev for their takes. (I'd invoke James McNellis, but he's moved on to smaller and more playful things at Roblox.) I don't feel qualified to sign off on my own.

**Comment by mstorsjo** [comment]
> Thanks for the proposal. As near as I can tell, the actual opcode is known as `pac_sign_return_address`, and the logic is largely as you've divined. There appears to be a flag that lets the signing/verifying opcodes be replaced with nops, so there's an additional bit of complexity involved somewhere. At least, I think so, if I'm reading things correctly, which is always a question. The signing step is "Step 1" in the internal comments, and saving registers is "Step 2", so we should probably renumber those and the following steps accordingly.

Thanks! I've updated the PR to use the correct name for the opcode, and renumbered the steps for the packed form.

File: docs/build/arm64-exception-handling.md [hunk 1]
```diff
@@ -313,6 +313,7 @@ The unwind codes are encoded according to the table below. All unwind codes are
 |  | 11101010: Custom stack for MSFT_OP_CONTEXT |
 |  | 11101100: Custom stack for MSFT_OP_CLEAR_UNWOUND_TO_CALL |
 |  | 1111xxxx: reserved |
+| `pac_sign_return_address` | 11111100: sign the return address in `x30` with `pacibsp` |
```
Comment (sigatrev) [inline docs/build/arm64-exception-handling.md:316]: `x30` is referred to as `lr` everywhere else in this document. We should use `lr` here was well for consistency.

Comment (mstorsjo) [reply docs/build/arm64-exception-handling.md]: Thanks - I updated the branch, fixing this issue now, so that it references `lr` instead.

**Review by sigatrev — APPROVED** [review]

**Comment by mstorsjo** [comment]
@sigatrev BTW, are you able to comment on what the Windows unwinder implementation does, if the actual code does use PAC (and running on HW that exposes it, with a Windows configuration that has taken it into use), but the unwind info is lacking these opcodes? (That's the current status quo if building code with Clang today, with `-mbranch-protection=standard`.)

**Comment by pmsjt** [comment]
> @sigatrev BTW, are you able to comment on what the Windows unwinder implementation does, if the actual code does use PAC (and running on HW that exposes it, with a Windows configuration that has taken it into use), but the unwind info is lacking these opcodes? (That's the current status quo if building code with Clang today, with `-mbranch-protection=standard`.) I.e., does the unwinder implicitly run `xpaci` on the return address after finishing `RtlVirtualUnwind`, or does it only do that (or similar) if explicitly told so by an unwinding opcode?

It must be explicitly present.

**Comment by mstorsjo** [comment]
> > @sigatrev BTW, are you able to comment on what the Windows unwinder implementation does, if the actual code does use PAC (and running on HW that exposes it, with a Windows configuration that has taken it into use), but the unwind info is lacking these opcodes? (That's the current status quo if building code with Clang today, with `-mbranch-protection=standard`.) I.e., does the unwinder implicitly run `xpaci` on the return address after finishing `RtlVirtualUnwind`, or does it only do that (or similar) if explicitly told so by an unwinding opcode?
> 
> It must be explicitly present.

Ok, thanks for the clarification!

**Review by corob-msft — APPROVED** [review]
@mstorsjo 
Thanks for the useful update.

**Comment by corob-msft** [comment]
#sign-off

**Pull request merged**